proposals/0000-context-fixes.rst

```diff
@@ -0,0 +1,176 @@
+.. proposal-number:: Leave blank. This will be filled in when the proposal is
+                     accepted.
+
+.. trac-ticket:: Leave blank. This will eventually be filled with the Trac
+                 ticket number which will track the progress of the
+                 implementation of the feature.
+
+.. implemented:: Leave blank. This will be filled in with the first GHC version which
+                 implements the described feature.
+
+.. highlight:: haskell
+
+This proposal is `discussed at this pull requst <https://github.com/ghc-proposals/ghc-proposals/pull/40>`_.
+
+Context fixes.
+==============
+
+I suggest to add a new syntactic form ``context`` which uniformly abstracts over a group of declarations. Consider this example, which does not do anything useful, but shows how the feature can be used to solve the configuration problem:
+
+::
+
+  context fixes progName in
+    foo :: Maybe Int -> Either String Int
+    foo Nothing  = Left $ progName ++ ": no number given"
+    foo (Just i) = bar i
+    
+    bar :: Int -> Either String Int
+    bar 0 = Left $ progName ++ ": zero no good"
+    bar n = Right $ n + 1
+
+
+This is equivalent to 
+
+::
+
+  foo :: String -> Maybe Int -> Either String Int
+  foo progName Nothing  = Left $ progName ++ ": no number given"
+  foo progName (Just i) = bar progName  i
+  
+  bar :: String -> Int -> Either String Int
+  bar progName  0 = Left $ progName ++ ": zero no good"
+  bar progName n = Right $ n + 1
+
+
+Motivation
+------------
+
+Haskell is good at abstracting. If I have a function like::
+
+  foo :: T -> S
+  foo t = … 42 …
+
+and I want to make this abstract in the number, I add a new parameter::
+
+  foo :: Int -> T -> S
+  foo n t = … n …
+
+Nice and easy. But what if I have a group of declarations like the following::
+
+  foo :: T -> S
+  foo t = … 42 … bar s … foo t …
+  
+  bar :: S -> T
+  bar s = … foo t … 42 …
+
+I not only have to add the parameter to every function involved, but I also have to change every function call within the group::
+
+  foo :: Int -> T -> S
+  foo n t = … n … bar n s … foo n t …
+  
+  bar :: Int -> S -> T
+  bar n s = … foo n t … n … 
+
+This works, but is clumpsy. Also, it is less easy to understand: As a reader of the new code it is not obvious to me that the `n` parameter is *the same* in all the invocations – after all, one of the calls between the functions could be `foo 42` or `foo (n+1)`.
+
+The suggested new language feature would allow me to express this as follows::
+
+  context fixes n in
+    foo :: T -> S
+    foo t = … n … bar s … foo t …
+
+    bar :: S -> T
+    bar s = … foo t … n …
+
+Note that:
+
+* Besides the indentation and replacing ``42`` by ``n``, I did not have to change any code.
+* It is very obvious to the reader that wihtin the indentet block, ``n`` is not changed.
+* The typo of ``foo`` is different within the scope of the ``context`` block: It is ``T -> S`` inside, but ``Int -> T -> S`` outside.
```

Comment:
Some typos in this paragraph: wihtin -> within, indentet -> indented, typo -> type?

And I would say "obvious" is enough :smile:

Some links to the background where you borrowed the syntax from would be useful (according to your Debian blog post Isabelle?).

What about going from "read only once at startup" to "read at configuration reloads"?